### ISSUE TITLE
Trouble working with large (2k points) sensor data in Rainbow Shading

### ISSUE THREAD

**Issue by IanMayo** [debrief/debrief#1808]
We allow an analyst to use rainbow shading to temporally navigate long blocks of sensor data.

An analyst had an ownship track plus a sensor with 1800 sensor cuts.  The sensor was rainbow-shaded.

The analyst selected some data, and created a new TMA leg from it.  Then he made that TMA leg the secondary track.

Then the Outline view (and the rest of Debrief) started losing icons, with mouse movement acting like an "eraser".

The error log shows an error "Graphic is disposed, being called from **Image.getImageData()**:
````
public ImageData getImageData() {
	if (isDisposed()) SWT.error(SWT.ERROR_GRAPHIC_DISPOSED);
````

My first suspicion is that RCP is "running out of handles" when it tries to allocate memory for the 1800 uniquely colored sensor icons.

If this is the case, the fix is to not show all those unique shades, but to show at most 100 shades.  If there are less than 100 sensor cuts, show the unique colors.  If there are more than 1000 sensor cuts, re-use each icon 10 times before generating the next one.  So the colors for the icons in the Outline view will be in groups rather than continuously shaded along the data.

I've had a look at the code. We use **CoreViewLabelProvider** to get images.  This does some processing, but significantly, the SWT color is retrieved from the **ColorHelper** class.  Hmm, I suggest that we create a Marker Interface for elements that should have "chunked" colors: **IChunkedColors**.  **SensorWrapper** would be given this marker interface.  Where the marker interface is found, we apply special formatting:
* if the element has a color that is a default Debrief color, let it use it
* if no, get the HSB value for the color (0..1).  Then chunk this to the 100 shades by multiplying by 100, taking the floor() value, then dividing by 100.  Then get this new color from the **ColorHelper**.

Note: in order to check if something is a Debrief color, we should probably have them available in an array/list somewhere.  Is it possible to create a COLORS static array object in the **DebriefColors** class?

The problem can be reproduced as follows:
* load **nonsuch.rep** - select **Dead Reckoning DR** as the import mode
* load **sensor.dsf** from the same folder. Follow the wizard, leaving **visible** tick box un-clicked
* open up the sensor data. 
* select all sensor cuts, right click and select **Keep starboard bearing**
* Right-click on sensor, and select **Sensor A / Resample data at  / 1 second**
* Right-click on sensor, and select **Shading / Shade in rainbow colors**
* Right-click on sensor, and select **Generate TMA solution for all cuts**
* Follow the wizard, put target at about 1nm range, speed at about 2 kts
* New TMA solution will appear in Outline view. Mark it a secondary

Around this point, the icons should start disappearing from the Outline View.  I can't reproduce this under Linux, the analyst experienced the problem under MS Windows.

**Note:** I've just reproduced the problem using MS Windows, roughly following the above script:

![rainbowicons](https://cloud.githubusercontent.com/assets/1108513/16268830/ce96f4e4-3887-11e6-8af2-0fdeae897144.PNG)

Note: I'm not fully convinced that I've diagnosed the issue correctly.  The fact that the issue doesn't arise when we apply rainbow shading, but once we mark the TMA solution as secondary suggests it may not be about colored sensor icons.

**Comment by IanMayo** [comment]
@theanuradha - do you have access to a MS Windows development machine?  I'm afraid this is a MSWin-specific issue.

**Comment by theanuradha** [comment]
@IanMayo I do have windows vm will try it there.

**Comment by theanuradha** [comment]
@IanMayo ON mac it seems fine too.  wil switch to windows ,(as it seems windows gc handler issue trigger this)

**Comment by theanuradha** [comment]
@IanMayo Ah on mac sensor cuts icons missing at start then I have to close and re open to get to them

**Comment by theanuradha** [comment]
![image](https://cloud.githubusercontent.com/assets/535713/16315031/2b110b80-399e-11e6-9ad7-9351525adb42.png)

**Comment by theanuradha** [comment]
so while it not completely mess Outline view it drop all icons it has on MAC. Maybe on linux try close and open  **Sensors** node see if get to this

**Comment by theanuradha** [comment]
![image](https://cloud.githubusercontent.com/assets/535713/16315418/d557d046-399f-11e6-84c2-d46ea53f6914.png)

**Comment by theanuradha** [comment]
@IanMayo  I have impl the IChunkedColors solution for Rainbow Shading. I have debug and verified this correctly  Chunked and skip shade for nex t100 elelents. Issue is on mac it still runs out of resource. will test on windows what happen . Please take look at changes on my branch I did not create PR yet. 
https://github.com/theanuradha/debrief/commit/ab704085c26a0ee4839a622ba0bd757a456b591c

**Comment by theanuradha** [comment]
@IanMayo  I added one more commit. Now on mAc it is not running out of resources but  colors look odd https://github.com/theanuradha/debrief/commit/8404e4eda424731e123941d5dc9002eddc4b34ca

![image](https://cloud.githubusercontent.com/assets/535713/16608103/34d1e91a-4366-11e6-8be5-6d876126a9ca.png)


maybe my **getChunkedColor** is wrong 
`public static java.awt.Color getChunkedColor(final java.awt.Color javaCol)
  {

    java.awt.Color theColor = javaCol;
    if (!Arrays.asList(DebriefColors.COLORS).contains(theColor))// TODO: Check for DebriefColors and
                                                                // ignore
    {
      float[] hsb =
          Color.RGBtoHSB(theColor.getRed(), theColor.getGreen(), theColor
              .getBlue(), null);

      // Chunked to 100 element shade
      hsb[0] = (float) (Math.floor(hsb[0] * 100) / 100);
      hsb[1] = (float) (Math.floor(hsb[1] * 100) / 100);
      hsb[2] = (float) (Math.floor(hsb[2] * 10000) / 100);
      return Color.getHSBColor(hsb[0], hsb[1], hsb[2]);
    }

    return javaCol;
  }`

**Comment by theanuradha** [comment]
@IanMayo sorry I think I found the issu it is typo in my code getChunkedColor let me fix and try

**Comment by theanuradha** [comment]
I have created the PR https://github.com/debrief/debrief/pull/1820. Will test On windows

**Comment by theanuradha** [comment]
@IanMayo  after I start review again I found the real issue .

Issue was really trigger by method **resetCacheFor**(tree) as it try to update images and clear cache at lower level and incorrectly dispose valid images . so app never ran out resources it just dispose incorrectly. I have rework **resetCacheFor(TreeViewer tree)**  method and even without **IChunkedColors** solution it works perfectly. I left the  **IChunkedColors** anyway please look at rework of **resetCacheFor** at https://github.com/debrief/debrief/pull/1820/commits/570699078b9d9e883146cff677c226cfbc503098

**Comment by IanMayo** [comment]
Fixed in #1820 

I've removed IChunkedColors, to try to keep the code as simple as possible.
